### PR TITLE
Keep Calypso alive while navigating to template parts

### LINKED ISSUE #37840: Back button navigation context is lost if you enter the template part editor.
In the FSE editor, the back button is overriden to take you to various pages in calypso, like themes or customer home depending which screen you entered the editor from.

However, as soon as you edit your header and footer, the back button context is lost. If you go back to the page editor, it will always then lead back to the all pages list instead of the place in calypso from which you entered the editor in the first place. This is because a full page reload is done when navigating to the template part editor in back. We need to make sure that we don't fully reload all of Calypso in that case, and just reload the iFrame.

### PR DESCRIPTION
**Pull request by noahtallen**
### Goals
1. Stop destroying the calypso navigation context when we navigate to/from the template part. This caused a bug where you would always be taken back to the "all pages" list, even if you started elsewhere, like customer home. This happened as long as you went into the template editor.
2. Improve performance. We are no longer loading all of calypso every time we transition between the page editor and template parts, just the stuff inside the iFrame.
3. A great accidental result here is that it should be more obvious what the code does, and simpler to work on in the future.

### Changes proposed in this Pull Request
* Add a `key` prop to the Iframe so that it resets if we start editing a different post.
* Remove code which updates the hrefs of the header/footer links.
* Remove code which does a server-side navigation if navigating to/from template pats.
* Introduce a handler which, on clicking "edit header/footer", sends a message to the Calypso code, which does a client-side navigation to/from the template part.

#### Testing instructions
1. Pull this and build it on calypso.localhost
2. Sync the latest and greatest FSE changes to your sandbox.
2. Sync the bridge server code (D35795-code) and sandbox `widgets.wp.com`
3. Navigate between the page editor and the template part editor and smoke test that everything works properly (especially making sure the URL in the browser nav bar matches the editor you are looking at).
4. Test creating pages/posts.
5. Test navigation in the post editor without FSE.

#### Test back button
1. From customer home on a site with FSE, try clicking "add a page"
Screen Shot 2019-11-22 at 2 34 53 PM

2. Choose a layout and publish the page
3. Edit the header.
4. Click the template part back button
5. Click the page back button. (it should be like "< Home" if you synced FSE)
6. You should end up at customer home.

Fixes #37840

### CODE REVIEW

**Comment by noahtallen** [comment]
Any ideas why the desktop tests are failing just for this PR?

Screen Shot 2019-11-19 at 9 42 25 PM

**Comment by Copons** [comment]
@noahtallen not sure if those errors are relevant, considering that desktop tests fail often to me for no reason, but a nifty trick to restart all the CI integrations is to remove and readd the Needs Review label. 🙂

**Comment by Copons** [comment]
This works as expected, but oh boy it's so hard to do a code review of the bridge! 😮 

I don't see anything offensive, but I'd love if others who were more involved with it than me could chime in.
cc @mmtr and @kwight who seem to appear often in the history of the file!

**Comment by gwwar** [comment]
> it's so hard to do a code review of the bridge

To make things easier for reviewers, it's usually nice to generate a phabricator diff for folks, then regenerate changes with master after files land in Calypso

**Comment by Copons** [comment]
> > it's so hard to do a code review of the bridge
> 
> To make things easier for reviewers, it's usually nice to generate a phabricator diff for folks, then regenerate changes with master after files land in Calypso

@gwwar Ah I meant "code reviews" literally.
Testing is a chore but the steps are straightforward enough (to me at least 😄).
But when it comes to _reading_ the bridge code, I'm totally lost and always have a very hard time following the "thread".

**Comment by noahtallen** [comment]
> restart all the CI integrations

I did restart the tests a few times :/

> This works as expected, but oh boy it's so hard to do a code review of the bridge!

I agree so much! Thankfully I wrote the code I'm getting rid of, which makes things easier :p

**Comment by noahtallen** [comment]
> generate a phabricator diff

Added D35795-code :)

> ok, those errors seem to be related

Yeah... even after a rebase it is still failing 🤔 I have no clue what to do about this. I guess I need to try to set up the desktop dev environment 🙃

**Comment by noahtallen** [comment]
Rebased everything on `master` including the diff. Testing should be easier now :)

**Comment by noahtallen** [comment]
Heck yeah, the tests pass!

File: apps/wpcom-block-editor/src/calypso/iframe-bridge-server.js [hunk 1]
```diff
@@ -642,66 +633,30 @@ function openCustomizer( calypsoPort ) {
...
+				payload: {
+					templatePartId,
+					unsavedChanges: select( 'core/editor' ).isEditedPostDirty(),
+				},
 			},
 			[ port2 ]
 		);
-	};
-
-	const unsubscribe = subscribe( () => {
```
Comment (vindl) [inline apps/wpcom-block-editor/src/calypso/iframe-bridge-server.js]: Nice LOC diff here! ✨ The removals make this code easier to grok.

**Pull request merged**